File: ignasr-app/src/main/java/limdongjin/ignasr/repository/IgniteRepositoryImpl.java
package limdongjin.ignasr.repository;

import org.apache.ignite.Ignition;
import org.apache.ignite.client.ClientCache;
import org.apache.ignite.client.IgniteClient;
import org.apache.ignite.configuration.ClientConfiguration;
import reactor.core.publisher.Mono;

import java.util.concurrent.CompletionStage;

public class IgniteRepositoryImpl implements IgniteRepository {
    private final ClientConfiguration clientConfiguration;
    public IgniteRepositoryImpl(ClientConfiguration clientConfiguration) {
        this.clientConfiguration = clientConfiguration;
    }

    public <K, V> Mono<Mono<K>> putAsync(String cacheName, K key, V value){
        Mono<IgniteClient> monoClient = Mono.just(clientConfiguration).map(Ignition::startClient);

        return monoClient.flatMap(client -> {
            CompletionStage<Mono<K>> monoCompletionStage = client
                    .getOrCreateCacheAsync(cacheName)
                     // :: IgniteClientFuture<ClientCache>.thenApplyAsync(cache -> monoK)
                    .thenApplyAsync(cache ->
                            Mono.fromCompletionStage(cache.putAsync(key, value).thenApply(vo -> key))
                                .doOnSuccess(unused -> client.close())
                    )
            ;
            return Mono.fromCompletionStage(monoCompletionStage);
        });
    }

    public <K, V> K put(String cacheName, K key, V value){
        try(var client = Ignition.startClient(clientConfiguration)){
            ClientCache<K, V> cache = client.getOrCreateCache(cacheName);
            cache.put(key, value);
            return key;
        }
    }

    public <K, V> V get(String cacheName, K key) {
        try(var client = Ignition.startClient(clientConfiguration)){
            ClientCache<K, V> cache = client.getOrCreateCache(cacheName);
            return cache.get(key);
        }
    }

    public <K, V> int size(String cacheName) {
        try(var client = Ignition.startClient(clientConfiguration)){
            ClientCache<K, V> cache = client.getOrCreateCache(cacheName);
            return cache.size();
        }
    }
}

File: ignasr-app-kt/src/main/kotlin/limdongjin/ignasr/repository/IgniteRepositoryImpl.kt
package limdongjin.ignasr.repository

import org.apache.ignite.Ignition
import org.apache.ignite.client.IgniteClient
import org.apache.ignite.configuration.ClientConfiguration
import reactor.core.publisher.Mono
import java.util.concurrent.CompletionStage


class IgniteRepositoryImpl(private val clientConfiguration: ClientConfiguration) : IgniteRepository {
    override fun <K: Any, V: Any> putAsync(cacheName: String, key: K, value: V): Mono<Mono<K>?> {
        val monoClient: Mono<IgniteClient> = Mono.just(clientConfiguration).map(Ignition::startClient)
        return monoClient.flatMap { client ->
            val monoCompletionStage: CompletionStage<Mono<K>> = client
                .getOrCreateCacheAsync<K, V>(cacheName)
                .thenApplyAsync { cache ->
                    Mono.fromCompletionStage(cache.putAsync(key, value).thenApply { _ -> key })
                        .doOnSuccess { _ -> client.close() }
                }
            Mono.fromCompletionStage(monoCompletionStage)
        }
    }

    override fun <K: Any, V: Any> put(cacheName: String, key: K, value: V): K {
        Ignition.startClient(clientConfiguration).use { client ->
            val cache = client.getOrCreateCache<K, V>(cacheName)
            cache.put(key, value)
            return key
        }
    }

    override operator fun <K: Any, V: Any> get(cacheName: String, key: K): V {
        Ignition.startClient(clientConfiguration).use { client ->
            val cache = client.getOrCreateCache<K, V>(cacheName)
            return cache[key]
        }
    }

    override fun <K: Any, V: Any> size(cacheName: String): Int {
        Ignition.startClient(clientConfiguration).use { client ->
            val cache = client.getOrCreateCache<K, V>(cacheName)
            return cache.size()
        }
    }
}